src/Microsoft.ML.Data/DataView/DetectAnomalyBySrCnnBatchTransform.cs

```diff
@@ -0,0 +1,332 @@
+                Contracts.Check(iinfo == 0);
+                return _outputColumnType;
+            }
+
+            // Get a predicate for the input columns.
+            public Func<int, bool> GetDependencies(Func<int, bool> predicate)
+            {
+                Contracts.AssertValue(predicate);
+
+                var active = new bool[Input.Count];
+                for (int col = 0; col < ColumnCount; col++)
+                {
+                    if (!predicate(col))
+                        continue;
+
+                    bool isSrc;
+                    int index = MapColumnIndex(out isSrc, col);
+                    if (isSrc)
+                        active[index] = true;
+                    else
+                        active[_inputColumnIndex] = true;
+                }
+
+                return col => 0 <= col && col < active.Length && active[col];
+            }
+        }
+
+        public bool CanShuffle => false;
+
+        public DataViewSchema Schema => SchemaBindings.AsSchema;
+
+        private readonly IDataView _source;
+        private readonly IHost _host;
+        private protected readonly Bindings SchemaBindings;
+        protected readonly string InputCol;
+
+        protected BatchTransformBase(IHostEnvironment env, IDataView input, string inputColumnName, string outputColumnName, DataViewType outputColumnType)
+        {
+            _host = env.Register("Batch");
+            _source = input;
+            SchemaBindings = new Bindings(input.Schema, inputColumnName, outputColumnName, outputColumnType);
+            InputCol = inputColumnName;
+        }
+
+        public long? GetRowCount() => _source.GetRowCount();
+
+        public DataViewRowCursor GetRowCursor(IEnumerable<DataViewSchema.Column> columnsNeeded, Random rand = null)
+        {
+            _host.CheckValue(columnsNeeded, nameof(columnsNeeded));
+            _host.CheckValueOrNull(rand);
+
+            var predicate = RowCursorUtils.FromColumnsToPredicate(columnsNeeded, SchemaBindings.AsSchema);
+
+            // If we aren't selecting any of the output columns, don't construct our cursor.
+            // Note that because we cannot support random due to the inherently
+            // stratified nature, neither can we allow the base data to be shuffled,
+            // even if it supports shuffling.
+            if (!SchemaBindings.AnyNewColumnsActive(predicate))
+            {
+                var activeInput = SchemaBindings.GetActiveInput(predicate);
+                var inputCursor = _source.GetRowCursor(_source.Schema.Where(c => activeInput[c.Index]), null);
+                return new BindingsWrappedRowCursor(_host, inputCursor, SchemaBindings);
+            }
+            var active = SchemaBindings.GetActive(predicate);
+            Contracts.Assert(active.Length == SchemaBindings.ColumnCount);
+
+            // REVIEW: We can get a different input predicate for the input cursor and for the lookahead cursor. The lookahead
```

Comment:
>REVIEW [](start = 15, length = 6)

@harishsk, do you think it is ok to leave this as a review comment, or should this be addressed?
To follow the suggestion in the comment we would need to introduce an abstract method that gets the active columns for the lookahead cursor - in the simple implementation it would simply return `_source.Schema[SchemaBindings._inputColumnIndex]`.

Reply:
Or even simpler: `_source.Schema[InputCol]`.

---
In reply to: [424217890](https://github.com/mengaims/machinelearning/pull/1#discussion_r424217890) [](ancestors = 424217890)